File: Creational/Singleton/SIngleton.ts
/**
 * Singleton logger
 */
class Logger {
  private logs: any[];
  private static instance: Logger;

  private constructor() {}

  public static getInstance(): Logger {
    if (!Logger.instance) {
      Logger.instance = new Logger();
    }

    return Logger.instance;
  }

  public debug(message: any) {
    this.logs.push(message);
    console.log(`[DEBUG]: ${message}`);
  }

  public getLogs(): any[] {
    return this.logs;
  }
}

class FirstImplementation {
  private logger: Logger;

  constructor() {
    this.logger = Logger.getInstance();
  }

  public displayMessage() {
    this.logger.debug('hello world from first.');
  }
}

class SecondImplementation {
  private logger: Logger;

  constructor() {
    this.logger = Logger.getInstance();
  }

  public displayMessage() {
    this.logger.debug('hello world from second.');
  }
}

let firstInstance = new FirstImplementation();
firstInstance.displayMessage();

let secondInstance = new SecondImplementation();
secondInstance.displayMessage();

let logger = Logger.getInstance();
console.log(logger.getLogs());

File: Creational/Singleton/Singleton.py
from __future__ import annotations


class Logger:
    _instances = {}
    _logs = []

    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            instance = super().__call__(*args, **kwargs)
            cls._instances[cls] = instance
        return cls._instances[cls]

    def debug(self, message) -> None:
        self._logs.append(message)
        print(f"[DEBUG]: {message}")

    def get_logs(self):
        return self._logs


class FirstImplementation:
    _logger: Logger

    def __init__(self) -> None:
        self._logger = Logger()

    def display_message(self) -> None:
        self._logger.debug("First implementations")


class SecondImplementation:
    _logger: Logger

    def __init__(self) -> None:
        self._logger = Logger()

    def display_message(self) -> None:
        self._logger.debug("Second implementations")


firstImplementation = FirstImplementation()
firstImplementation.display_message()

secondImplementation = SecondImplementation()
secondImplementation.display_message()


logger = Logger()
print(logger.get_logs())
